### ISSUE TITLE
Add abillity for braille to be tethered to speech

### ISSUE THREAD

**Issue by Emil-18** [nvaccess/nvda#15898]
### Is your feature request related to a problem? Please describe.
As some features in NVDA and add-ons don't work with braille, the braille display should be able to show what NVDA is speaking to give users who only use braille access to the same functions
### Describe the solution you'd like
A new tether option, that tethers the braille display to speech.

**Comment by Adriani90** [comment]
How should this work? Speech is actually tetered to a certain cursor, i.e. to system cursor, system caret, mouse cursor or review cursor / navigator.
Braille is already supporting this as well.
What exactly would you like to have shown in braille?

**Comment by Emil-18** [comment]
When braille is tethered to speech, I would like speech.speech.speak to join every string in speechSequence with a space between them, and create a text region with it as an argument. I think this text region should be appended to a list containing everything NVDA has spoken since speech was last canceled, and then show this list of regions on the braille display in the main buffer. Hope this makes sence.

**Comment by josephsl** [comment]
Hi,

If I understand you correctly, you would like something similar to speech history? If I remember right, JAWS's speech history (formerly called speech box) feature shows the last spoken utterance on the braille display, not necessarily a history of speech up to a certain point.

Thanks.

**Comment by Emil-18** [comment]
@josephsl More like a speech replacement for braille users, but i guess speech history also works. I could make a pull request to show what I mean, but aren't I supposed to wait until the issue is triaged before making a pull request?

**Comment by Adriani90** [comment]
You can do a pull request even without creating an issue at all. Triage does no affect the process of creating pull requests.Von meinem iPhone gesendetAm 10.12.2023 um 21:42 schrieb Emil-18 ***@***.***>:﻿
@josephsl More like a speech replacement for braille users, but i guess speech history also works. I could make a pull request to show what I mean, but aren't I supposed to wait until the issue is triaged before making a pull request?

**Comment by Adriani90** [comment]
Did you test this already? I imagine such things like "linefeed" or "space" or any other punctuation is completely displayed on the braille display when it is tetered to speech. Correct?
Is this really practical for a braille user to have these things displayed completely?

**Comment by josephsl** [comment]
Hi, I get a feeling that this issue is a duplicate of another one. Thanks.

**Comment by seanbudd** [comment]
>You can do a pull request even without creating an issue at all. Triage does no affect the process of creating pull requests.

While you can do this, it's not recommended.
Pull requests are a lot of work, it's not worth starting one if a feature suggestion gets rejected, or further advice is provided by NV Access.

See [contributing docs](https://github.com/nvaccess/nvda/blob/master/projectDocs%2Fdev%2Fcontributing.md)

**Comment by LeonarddeR** [comment]
> You can do a pull request even without creating an issue at all.

While this is technically possible, it is generally discouraged unless it is very obvious from the pr that it fixes an issue that needs fixing.

In this case, I guess it is the add-ons primary responsibility to respect braille users. That said, I think the current proposal makes sense. It will however make the options at this point even more complex.

**Comment by CyrilleB79** [comment]
As explained before speech is not a specific part of the system, nor a precise location, on contrary to system or review cursors.

Thus, the concept of tethering to speech may be confusing from a UX point of view.

Why not create a completely distinct option: "Braille only logs the speech" which could be enabled or disabled. When this option is enabled, various items of the braille settings panel do not apply and should be disabled, e.g. Tether to, message displaying, display selection, etc.
A better name than my suggestion should be found though.

**Comment by seanbudd** [comment]
@CyrilleB79's proposal makes sense

**Comment by ruifontes** [comment]
I agree with @CyrilleB79.
BrailledSpeech? It will show ecerything spoken by NVDA in the Braille display?
We will have the possibility to advance, go forward, select and so on?

**Comment by Emil-18** [comment]
After thinking about it, I agree as well with @CyrilleB79, as it would be a compleatly different way of using braille. It would also allow for adding more braille modes in the future

**Comment by Emil-18** [comment]
I have started working on a PR that fixes this issue. Is it OK if I call the speech mode speech emulation, and the normal mode cursor tethering?

**Comment by seanbudd** [comment]
I think "speech mode emulation" is good - particularly if it mirrors the current speech mode's behaviour - i.e. beeps, talk, on-demand, none.
Normal mode I think could be improved - perhaps "review mode" as braille behaviour will instead match the review mode, not the speech mode.

**Comment by zersiax** [comment]
There are various instances of speech firing, but braille not firing. Until recently, aria-live was a common one, but various things like games that talk to the NVDAController will do this as well, so I can see this opening doors for braille-only users. Just an FYI.

**Comment by Emil-18** [comment]
@seanbudd I don't plan for the speech emulation mode to follow the speech mode. I think it is better if it shows what NVDA would have said if it was in speech mode, regardless of what speech mode the user is actualy in. In this way, the user can turn off speech with NVDA+s, and still read what NVDA would have said. What about calling the normal mode, "follow cursors", as I feal "follow review" is a bit misleading, as you can tether braille to review as well?

**Comment by CyrilleB79** [comment]
"Follow cursor" for the normal braille mode is much better than "Follow review mode" which is misleading with the concept of tethering to review.

For the new mode, I would call it "Speech transcription" (better); should be clear with context. Or "Transcription of the speech in braille".

Thus I suggest:
* either a checkbox, disabled by default, named "Enable speech transcription" or "Transcribe the speech in braille"
* or for more clarity, a combo-box with the two following options:
  * Follow cursor (option selected by default)
  * Speech transcription

There is no need to have a feature flag for this because I do not expect the normal mode (follow cursor) to change.

The braille options should be reordered in the panel, so that all the options which do not apply to the new mode follow each other. More specifically, I guess that there is no concept of cursor nor selection in this mode. Is it correct?

@Emil-18, you should probably list here separately in two lists the options that apply in any mode (e.g. braille tables) and the options that apply only in braille mode = follow cursor (e.g. tether to), so that everyone agree with it.

At last, I agree, if possible, that the speech transcription mode should still operate even if speech mode is off since there is definitely a use case for it. However, it's a "nice to have", not a "must have", i.e. if it's too difficult from a technical point of view, this requirement should not block the development of this feature.

**Comment by Emil-18** [comment]
The options that I think only apply in cursor following mode is:
* Expand to computer braille for the word at the cursor
* Show cursor
* Blink cursor
* Cursor blink rate
* Cursor shape for focus
* Cursor shape for review
* Show messages
* Message timeout
* Tether Braille
* Move system caret when routing review cursor
* Read by paragraph
* Focus context presentation
* Show selection
@CyrilleB79 There is no concept of cursor or selection, as this information will be reported in the same way as in speech. What do you think is better by calling it speech transcription instead of speech emulation? My idea is that you would use this mode because you are unable to use speech for whatever reason, therefor emulating using speech via this mode

**Comment by CyrilleB79** [comment]
I agree with the options that you have listed as applying only in cursor following mode.
I think that it would make sense to reorder the options so that all the options that apply only in cursor following mode are grouped together, I mean, following each other in the GUI.

We may also wonder if grouping can be used to organize a bit more this panel, as done in Document formatting settings or Advanced settings. But I have not checked if it's easy to define interesting groups.

To be clear, the remaining options that apply in all mode (cursor following and speech transcription/emulation) are the following:
* Braille display
* Output table
* Input table
* Avoid splitting words when possible
* Interrupt speech while scrolling

>       @CyrilleB79 There is no concept of cursor or selection, as this information will be reported in the same way as in speech.

OK. Makes sense. It was just to confirm this.

Another question:
If you press a routing cursor, I'd imagine that there is no action. Do you confirm this?

> What do you think is better by calling it speech transcription instead of speech emulation? My idea is that you would use this mode because you are unable to use speech for whatever reason, therefor emulating using speech via this mode

"Transcription" is used, at least in French, in situations where what is spoken should be written, e.g. live caption of videos, report of a trial in the field of justice, etc. "Emulation" let me think of a process allowing to get the same look and feel or the same effect as the original thing, e.g. emulation of an old game console on a computer, emulate key press, etc. Hence my proposal.

Anyway if you and other prefer "Emulation", it's OK for me.

**Comment by josephsl** [comment]
Hi,

At least in English, I think the most descriptive terms could be "presentation" or "history." Transcription, at least in English (and some parts of the world) involve documentation or writing messages from one form to another. However, what's really being asked for is a way to present spoken content in braille, similar to what happens with ui.message function.

One example is "focus context presentation" combo box. This conbo box controls how focus ancestry info is presented in braille. I think we should call the new option something similar to this.

As for "follow cursor" discussion: I think "cursor trakincg" might be a better term since braille might "track" spoken content based on the word "tethering".

Thanks.

**Comment by Emil-18** [comment]
@CyrilleB79 I don't know what the routing action should have bin in speech mode, so no. I think the settings you listed are relevent in both modes. It is the same for me if the speech mode is called speech transcription or speech emulation. I think the settings that are only relevent to the cursor following mode should be in a group. In this way, it is much less error prone to disable/enable them when toggeling the braille mode setting, as then we can just disable/enable the entire group
